src/sagas/delete-project.saga.js

```diff
@@ -0,0 +1,96 @@
+// @flow
+import { remote } from 'electron';
+import { call, put, select, takeEvery } from 'redux-saga/effects';
+
+import {
+  SHOW_DELETE_PROJECT_PROMPT,
+  finishDeletingProjectFromDisk,
+  selectProject,
+  createNewProjectStart,
+} from '../actions';
+import { getProjectsArray } from '../reducers/projects.reducer';
+
+import type { Action } from 'redux';
+import type { Saga } from 'redux-saga';
+import type { Project } from '../types';
+
+const { dialog, shell } = remote;
+
+export const getNextProjectId = (
+  projects: Array<Project>,
+  idToDelete: string
+) => {
+  const projectIndex = projects.findIndex(project => project.id === idToDelete);
+
+  if (projectIndex === -1) {
+    throw new Error(
+      "The project you're attempting to delete does not appear to exist within the known projects. We're not sure how this is possible :/"
+    );
+  }
+
+  // If we're about to delete the last project in the list, there is no other
+  // project to select.
+  if (projects.length === 1) {
+    return null;
+  }
+
+  // We want to select the next project in the list after the to-be-deleted one.
+  // If the to-be-deleted project IS last in the list, select the one before.
+  const isLastProjectBeingDeleted = projectIndex === projects.length - 1;
+
+  const nextProject = isLastProjectBeingDeleted
+    ? projects[projectIndex - 1]
+    : projects[projectIndex + 1];
+
+  return nextProject.id;
+};
+
+export function* deleteProject({ project }: Action): Saga<void> {
+  const response = yield call(dialog.showMessageBox, {
```

Comment:
@mik639 and I had a rather lengthy discussion about this and I'm still not sure of the outcome. So there are 3 different ways to write line 49:

```js
const response = yield call(dialog.showMessageBox, {
const response = yield call([dialog, 'showMessageBox'], {
const response = yield call([dialog, dialog.showMessageBox], {
```

If I remember correctly, only the third form passed flow typing, which is added in #154. IMO the first form is by far the most intuitive, but flow typing says otherwise.

Reply:
@superhawk610 yes, the first one is the most intuitive one. Looking at the [typed-def](https://github.com/flow-typed/flow-typed/blob/0dda5b292f63738b8b6d2c46db1477ffa6dea97c/definitions/npm/redux-saga_v0.16.x/flow_v0.56.0-/redux-saga_v0.16.x.js#L401) I couldn't see how we get the typing right - really hard to read.

Have you seen this [issue](https://github.com/redux-saga/redux-saga/issues/864)? Not sure if there is something that's helping.

We should try to get the intuitive syntax to work with flow - I think Sagas are complicated enough even with-out this. (I'll have to learn how Sagas are working - I haven't used them before.)